Ginger/GingerCoreNET/Drivers/CoreDrivers/Mobile/Appium/GenericAppiumDriver.cs

```diff
@@ -1614,33 +1714,257 @@ public bool IsElementObjectValid(object obj)
+            }
+            finally
+            {
+                mIsDriverBusy = false;
+            }
+        }
+
+        public object LocateElementByLocators(ObservableList<ElementLocator> Locators)
+        {
+            object elem = null;
+            foreach (ElementLocator locator in Locators)
+            {
+                locator.StatusError = string.Empty;
+                locator.LocateStatus = ElementLocator.eLocateStatus.Pending;
+            }
+
+            foreach (ElementLocator locator in Locators.Where(x => x.Active == true).ToList())
+            {
+                if (!locator.IsAutoLearned)
+                {
+                    elem = LocateElementIfNotAutoLeared(locator);
+                }
+                else
+                {
+                    elem = LocateElementByLocator(locator);
+                }
+
+                if (elem != null)
+                {
+                    locator.StatusError = string.Empty;
+                    locator.LocateStatus = ElementLocator.eLocateStatus.Passed;
+                    return elem;
+                }
+                else
+                {
+                    locator.LocateStatus = ElementLocator.eLocateStatus.Failed;
+                }
+            }
+
+            return null;
+        }
+
+        public object LocateElementByLocator(ElementLocator EL)
+        {
+            //need to override regular selenium driver locator if needed, 
+            //if not then to run the regular selenium driver locator for it to avoid duplication   
+            object elem;
+
+            try
+            {
+                switch (EL.LocateBy)
+                {
+                    case eLocateBy.ByResourceID:
+                        elem = Driver.FindElementById(EL.LocateValue);
+                        break;
+
+                    case eLocateBy.ByRelXPath:
+                    case eLocateBy.ByXPath:
+                        elem = Driver.FindElementByXPath(EL.LocateValue);
+                        break;
+
+                    default:
+                        elem = mSeleniumDriver.LocateElementByLocator(EL);
+                        break;
+                }
+            }
+            catch(Exception exc)
+            {
+                elem = null;
+                EL.StatusError = exc.Message;
+                EL.LocateStatus = ElementLocator.eLocateStatus.Failed;
+            }
+
+            return elem;
+        }
+
+        private object LocateElementIfNotAutoLeared(ElementLocator el)
+        {
+            ElementLocator evaluatedLocator = el.CreateInstance() as ElementLocator;
+            //ValueExpression VE = new ValueExpression(this.Environment, this.BusinessFlow);
+            //evaluatedLocator.LocateValue = VE.Calculate(evaluatedLocator.LocateValue);
+            return LocateElementByLocator(evaluatedLocator);
         }
 
         public ElementInfo GetMatchingElement(ElementInfo latestElement, ObservableList<ElementInfo> originalElements)
```

Comment:
does all the method added for learn and update are different from methods placed in Selenium driver?